vm/natives.c

```diff
@@ -1138,5 +1138,393 @@ const AsebaNativeFunctionDescription AsebaNativeDescription_rand =
 	}
 };
 
+// standard native functions for deque
+
+// A deque of capacity K is stored in an array of size K+2 that contains N, J, and a
+// gap buffer [Finseth 1991]. Operations on the deque may use an index I relative to
+// the front of the deque. For example a deque of capacity 8 containing N=6 elements
+// stored starting at position J=3, and an index I=2 pointing before the third element:
+// +---+---+---+---+---+---+---+---+---+
+// + 6 | 3 | 5 |   |   | 0 | 1 | 2 | 4 |
+// +---+---+---+---+---+---+---+---+---+
+//   N   J              *J      *I
+// Deque operations that accept another array as an addend will treat it as a tuple of
+// elements, to be inserted, copied, or set as a unit. The push_* and pop_* operations
+// respect the size of the addend, and a deque can get or pop a tuple into an interval
+// in another array. For example, call.pop_front(result[3:5], dq) will pop three elements
+// from a deque stored in array dq into positions 3:5 of the array result. It is the
+// programmer's responsibility to use consistent tuple sizes.
+
+// Inserting and erasing must shift existing elements in the deque. The deque_shift private
+// function replaces the followinf specific loops:
+//
+// Insert in left half: shift prefix elements left
+//		for (k = 0; k <= index_val - 1 + src_length - 1; k++)
+//			vm->variables[dest + 2 + ((dq_start + k) % dq_capacity)] = vm->variables[dest + 2 + ((dq_start + k + src_length) % dq_capacity)];
+// Insert in right half: shift prefix elements left
+//		for (k = dq_size + src_length - 1; k > index_val + src_length - 1; k--)
+//			vm->variables[dest + 2 + ((dq_start + k) % dq_capacity)] = vm->variables[dest + 2 + ((dq_start + k - src_length) % dq_capacity)];
+// Erase in left half: shift prefix elements right
+//		for (k = index_val; k > 0; k--)
+//			vm->variables[dest + 2 + ((dq_start + k) % dq_capacity)] = vm->variables[dest + 2 + ((dq_start + k - 1) % dq_capacity)];
+// Erase in right half: shift prefix elements left
+//		for (k = index_val; k < dq_size - 1; k++)
+//			vm->variables[dest + 2 + ((dq_start + k) % dq_capacity)] = vm->variables[dest + 2 + ((dq_start + k + 1) % dq_capacity)];
+
+static void deque_shift(AsebaVMState *vm, uint16 dq, uint16 dq_capacity, uint16 target, sint16 last, sint16 delta)
+{
+	for ( ; (delta < 0 ? target >= last : target <= last) ; (delta < 0 ? target-- : target++) )
+		vm->variables[dq + 2 + (target % dq_capacity)] = vm->variables[dq + 2 + ((target + delta) % dq_capacity)];
+}
+
+void AsebaNative_deqsize(AsebaVMState *vm)
+{
+	// variable pos
+	uint16 size = AsebaNativePopArg(vm);
```

Comment:
It is a very good point. Until now the prefixes in Aseba stdlib are conceptually modules, such as "math.*". In the case of deque, we can continue to see it as a module, in case we should have destination first, or see it as a pseudo-object, in case we should have the _this_ pointer first.

It is true that personally I like the pseudo object, and that on the one hand it could be interesting to teach this concept ; but in the other hand, maybe it is adding a new convention for a special case, which is bad. Maybe the best is to write down the different options, look at them, and take the one that _feels_ the right one?